migration/migration.go

```
@@ -1,25 +1,130 @@
+// in -1 + 1 = 0 which is exactly what we want as the first version.
+func getCurrentVersion(db *gorm.DB) (int, error) {
+	if !db.HasTable("version") {
+		return -1, nil
+	}
+
+	res := db.Raw("select max(version) from version")
+	if res.Error != nil {
+		return -1, res.Error
 	}
 
+	type VersionResult struct {
+		version int
+	}
+	var v VersionResult
+	res.Scan(&v)
+	return v.version, res.Error
+}
+
+// Func defines the type of function that can be part of a migration sequence
+type Func func(tx *gorm.DB) error
+
+// executeSQLFile loads the given filename from the packaged SQL files and
+// executes it on the given database
+func executeSQLFile(filename string) Func {
+	return func(db *gorm.DB) error {
+		data, err := Asset(filename)
+		if err != nil {
+			return err
+		}
+		return db.Exec(string(data)).Error
+	}
+}
+
+// Migrate executes the required migration of the database on startup
+func Migrate(db *gorm.DB) error {
+
+	migrations := [][]Func{}
+
+	// Version 0
+	migrations = append(migrations, []Func{executeSQLFile("000-bootstrap.sql")})
+
+	// Version N
+	//
+	// In order to add an upgrade, simply append an array of MigrationFunc to the
+	// the end of the "migrations" slice. The version numbers are determined by
+	// the index in the array. The following code in comments show how you can
+	// do a migration in 3 steps. If one of the steps fails, the others are not
+	// executed.
+	// If something goes wrong during the migration, all you need to do is return
+	// an error that is not nil.
+
+	/*
+		migrations = append(migrations, []Func{
+			func(db *gorm.DB) error {
+				// Execute random go code
+				return nil
+			},
+			executeSQLFile("YOUR_OWN_FILE.sql"),
+			func(db *gorm.DB) error {
+				// Execute random go code
+				return nil
+			},
+		})
+	*/
+
+	ts := models.NewGormTransactionSupport(db)
+	var err error
+	for nextVersion := 0; nextVersion < len(migrations) && err == nil; nextVersion++ {
+
+		err = transaction.Do(ts, func() error {
```

The current impl seems a bit off:

Loop over all possible migrations
  Get current version, Set nextversion
    fail to execute current version
    move on to next version
    new current version works ok, previous failure is lost

?